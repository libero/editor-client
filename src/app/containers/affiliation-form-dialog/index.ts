export * from './affiliation-form-dialog';
export * from './connected-affiliation-form-dialog';
