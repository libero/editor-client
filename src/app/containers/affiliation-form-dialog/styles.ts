import { makeStyles } from '@material-ui/core/styles';

export const useAffiliationFormStyles = makeStyles((theme) => ({
  root: {},

  inputField: {
    marginBottom: theme.spacing(3)
  },

  buttonPanel: {
    display: 'flex',
    marginTop: theme.spacing(1),
    '& > *:last-child': {
      marginLeft: theme.spacing(2)
    }
  },

  spacer: {
    flexGrow: 1
  }
}));
