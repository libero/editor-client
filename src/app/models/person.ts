import { EditorState } from 'prosemirror-state';
import { DOMParser as ProseMirrorDOMParser } from 'prosemirror-model';
import { gapCursor } from 'prosemirror-gapcursor';
import { history } from 'prosemirror-history';
import { dropCursor } from 'prosemirror-dropcursor';
import { keymap } from 'prosemirror-keymap';
import { baseKeymap } from 'prosemirror-commands';

import * as bioConfig from 'app/models/config/author-bio.config';
import { Affiliation } from 'app/models/affiliation';
import { getTextContentFromPath, makeSchemaFromConfig } from 'app/models/utils';
import { buildInputRules } from 'app/models/plugins/input-rules';
import { BackmatterEntity } from 'app/models/backmatter-entity';
import { JSONObject } from 'app/types/utility.types';

export class Person extends BackmatterEntity {
  firstName: string;
  lastName: string;
  suffix?: string;
  isAuthenticated?: boolean;
  orcid?: string;
  email?: string;
  bio?: EditorState;
  isCorrespondingAuthor?: boolean;
  affiliations?: string[];
  hasCompetingInterest?: boolean;
  competingInterestStatement?: string;

  constructor(data?: JSONObject | Element, notesXml?: Element) {
    super();
    this.createEntity(data);
    if (data instanceof Element && notesXml instanceof Element) {
      this.setCompetingInterests(data, notesXml);
    }
  }

  public getDisplayName(): string {
    return [this.firstName, this.lastName, this.suffix].filter((_) => _).join(' ');
  }

  public getAffiliationsLabels(affiliations: Affiliation[]): string[] {
    return this.affiliations
      .map((affiliationId) => {
        const affiliation = affiliations.find(({ id }) => id === affiliationId);
        return affiliation ? affiliation.label : undefined;
      })
      .filter(Boolean)
      .sort((a, b) => Number(a) - Number(b));
  }

  clone(): Person {
    const person = new Person();
    person.firstName = this.firstName;
    person.lastName = this.lastName;
    person.suffix = this.suffix;
    person.isAuthenticated = this.isAuthenticated;
    person.email = this.email;
    person.bio = this.bio;
    person.isCorrespondingAuthor = this.isCorrespondingAuthor;
    person.affiliations = [...this.affiliations];
    person.hasCompetingInterest = this.hasCompetingInterest;
    person.competingInterestStatement = this.competingInterestStatement;

    return person;
  }

  protected fromXML(xml: Element): void {
    const orcidEl = xml.querySelector('contrib-id[contrib-id-type="orcid"]');

    this._id = xml.getAttribute('id') || this._id;
    this.firstName = getTextContentFromPath(xml, 'name > given-names');
    this.lastName = getTextContentFromPath(xml, 'name > surname');
    this.suffix = getTextContentFromPath(xml, 'name > suffix');
    this.isAuthenticated = orcidEl ? orcidEl.getAttribute('authenticated') === 'true' : false;
    this.orcid = orcidEl ? this.getOrcid(orcidEl.textContent) : '';
    this.bio = this.createBioEditorStateFromXml(xml.querySelector('bio'));
    this.email = getTextContentFromPath(xml, 'email');
    this.isCorrespondingAuthor = xml.getAttribute('corresp') === 'yes';
    this.affiliations = Array.from(xml.querySelectorAll('xref[ref-type="aff"]')).map((xRef) =>
      xRef.getAttribute('rid')
    );
  }

  protected fromJSON(json: JSONObject): void {
    this._id = (json.id as string) || this._id;
    this.firstName = json.firstName as string;
    this.lastName = json.lastName as string;
    this.suffix = json.suffix as string;
    this.isAuthenticated = json.isAuthenticated as boolean;
    this.orcid = json.orcid as string;
    this.bio = json.bio
      ? this.createBioEditorStateFromJSON(json.bio as JSONObject)
      : this.createBioEditorStateFromXml();
    this.email = json.email as string;
    this.isCorrespondingAuthor = json.isCorrespondingAuthor as boolean;
    this.affiliations = (json.affiliations as string[]) || [];
  }

  protected createBlank(): void {
    this.firstName = '';
    this.lastName = '';
    this.suffix = '';
    this.isAuthenticated = false;
    this.orcid = '';
    this.bio = this.createBioEditorStateFromXml();
    this.email = '';
    this.isCorrespondingAuthor = false;
    this.affiliations = [];
  }

  private getOrcid(orcidUrl: string): string {
    const matches = orcidUrl.match(/(([0-9]{4}-?){4})/g);
    if (matches) {
      return matches[0];
    }
    return '';
  }

  private setCompetingInterests(dataXml: Element, notesXml: Element): void {
    const competingInterestEl = Array.from(notesXml.querySelectorAll('[fn-type="COI-statement"]')).find(
      (fnEl: Element) => {
        const id = fnEl.getAttribute('id');
        return dataXml.querySelector(`xref[ref-type="fn"][rid="${id}"]`);
      }
    );

    this.hasCompetingInterest = competingInterestEl
      ? competingInterestEl.textContent !== 'No competing interests declared'
      : false;
    this.competingInterestStatement = competingInterestEl ? competingInterestEl.textContent : '';
  }

  private createBioEditorStateFromXml(bio?: Element): EditorState {
    const schema = makeSchemaFromConfig(bioConfig.topNode, bioConfig.nodes, bioConfig.marks);
    return EditorState.create({
      doc: bio ? ProseMirrorDOMParser.fromSchema(schema).parse(bio) : undefined,
      schema,
      plugins: [buildInputRules(), gapCursor(), dropCursor(), keymap(baseKeymap), history()]
    });
  }

  private createBioEditorStateFromJSON(json: JSONObject): EditorState {
    const blankState = this.createBioEditorStateFromXml();
    return EditorState.fromJSON(
      {
        schema: blankState.schema,
        plugins: blankState.plugins
      },
      json
    );
  }
}

export function createAuthorsState(authorsXml: Element[], notesXml: Element | undefined): Person[] {
  return authorsXml.map((xml) => new Person(xml, notesXml));
}
