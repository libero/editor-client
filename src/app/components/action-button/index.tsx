import React from 'react';
import { Button, PropTypes } from '@material-ui/core';
import { mapValues } from 'lodash';
import AddIcon from '@material-ui/icons/Add';

import { useActionButtonStyles } from './styles';

type MuiButtonVariant = 'text' | 'outlined' | 'contained';

const VARIANTS = {
  containedWarning: {
    variant: 'contained' as MuiButtonVariant,
    classes: {
      root: 'containedRedButton'
    }
  },
  outlinedWarning: {
    variant: 'outlined' as MuiButtonVariant,
    classes: {
      root: 'outlinedRedButton'
    }
  },
  primaryContained: {
    variant: 'contained' as MuiButtonVariant,
    color: 'primary',
    classes: {}
  },
  secondaryOutlined: {
    variant: 'outlined',
    classes: {}
  },
  addEntity: {
    size: 'small',
    classes: { root: 'addEntityButton_root', startIcon: 'addEntityButton_icon' },
    variant: 'text',
    startIcon: <AddIcon />
  }
} as {
  [key: string]: {
    variant: MuiButtonVariant;
    classes: Record<string, string>;
    color?: PropTypes.Color;
  };
};

export type VariantType = keyof typeof VARIANTS;

interface ActionButtonProps {
  onClick: () => void;
  title: string;
  variant: VariantType;
  className?: string;
}

export const ActionButton: React.FC<ActionButtonProps> = (props) => {
  const classes = useActionButtonStyles();
  const variantProps = VARIANTS[props.variant];
  const variantClasses = mapValues(variantProps.classes, (className: string) => classes[className]);
  const buttonProps = { ...variantProps, classes: variantClasses };
  return (
    <Button onClick={props.onClick} {...buttonProps} className={props.className}>
      {props.title}
    </Button>
  );
};
