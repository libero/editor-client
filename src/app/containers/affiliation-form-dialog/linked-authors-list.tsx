import React, { useState, useCallback } from 'react';
import DeleteIcon from '@material-ui/icons/Delete';
import { get } from 'lodash';
import { FormControl, InputLabel, MenuItem, Select, IconButton } from '@material-ui/core';

import { getAuthorDisplayName, Person } from 'app/models/person';
import { useAffiliationFormStyles } from './styles';
import { ActionButton } from 'app/components/action-button';

interface LinkedAuthorsListProps {
  linkedAuthors: Person[];
  allAuthors: Person[];
  onChange: (selkectedAuthors: Person[]) => void;
}

const renderAuthorSelectListItem = (author: Person) => (
  <MenuItem key={author.id} value={author.id}>
    {getAuthorDisplayName(author)}
  </MenuItem>
);

export const LinkedAuthorsList: React.FC<LinkedAuthorsListProps> = ({ linkedAuthors, allAuthors, onChange }) => {
  const [userLinkedAuthors, setUserLinkedAuthors] = useState<Person[]>(
    linkedAuthors.length > 0 ? linkedAuthors : [null]
  );

  const triggerOnChange = useCallback(
    (authorsList) => {
      onChange(authorsList.filter((author) => Boolean(author)));
    },
    [onChange]
  );

  const getAuthorSelectList = useCallback(
    (selectedAuthor) => {
      const selectedAuthorId = get(selectedAuthor, 'id');
      const linkedIds = new Set(userLinkedAuthors.map((author: Person) => get(author, 'id')));
      return allAuthors.filter(({ id }) => !linkedIds.has(id) || selectedAuthorId === id);
    },
    [userLinkedAuthors, allAuthors]
  );

  const deleteRow = useCallback(
    (index) => () => {
      const updatedList = [...userLinkedAuthors];
      updatedList.splice(index, 1);
      setUserLinkedAuthors(updatedList);
      triggerOnChange(updatedList);
    },
    [userLinkedAuthors, setUserLinkedAuthors, triggerOnChange]
  );

  const updateRow = useCallback(
    (index) => (event) => {
      const updatedList = [...userLinkedAuthors];
      updatedList[index] = allAuthors.find(({ id }) => id === event.target.value);
      setUserLinkedAuthors(updatedList);
      triggerOnChange(updatedList);
    },
    [userLinkedAuthors, allAuthors, triggerOnChange]
  );

  const addEmptyRow = useCallback(() => {
    setUserLinkedAuthors([...userLinkedAuthors, null]);
  }, [userLinkedAuthors, setUserLinkedAuthors]);

  const classes = useAffiliationFormStyles();
  return (
    <div className={classes.inputField}>
      {userLinkedAuthors.map((author: Person, index: number) => (
        <div className={classes.affiliatedAuthorRow} key={author?.id || index}>
          <FormControl key={get(author, 'id', index)} variant="outlined" className={classes.affiliatedAuthorInput}>
            <InputLabel shrink id="author-affiliations-label">
              Affiliated Author
            </InputLabel>
            <Select
              labelId="author-affiliations-label"
              displayEmpty
              value={get(author, 'id', '')}
              onChange={updateRow(index)}
              label="Affiliated Author"
            >
              <MenuItem value={''}>Select affiliated author</MenuItem>
              {getAuthorSelectList(author).map(renderAuthorSelectListItem)}
            </Select>
          </FormControl>
          <IconButton
            classes={{ root: classes.deleteButton }}
            onClick={deleteRow(index)}
            disabled={userLinkedAuthors.length <= 1}
          >
            <DeleteIcon fontSize="small" />
          </IconButton>
        </div>
      ))}
      <ActionButton variant="addEntity" title="Author" onClick={addEmptyRow} />
    </div>
  );
};
