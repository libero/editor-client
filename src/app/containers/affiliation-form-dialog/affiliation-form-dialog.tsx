import React, { SyntheticEvent, useCallback, useState } from 'react';
import { TextField } from '@material-ui/core';
import { useSelector } from 'react-redux';
import { set, cloneDeep } from 'lodash';

import { useAffiliationFormStyles } from './styles';
import { renderConfirmDialog } from 'app/components/prompt-dialog';
import { ActionButton } from 'app/components/action-button';
import { Affiliation, createAffiliation } from 'app/models/affiliation';
import { getAffiliatedAuthors, getAuthors } from 'app/selectors/manuscript.selectors';
import { LinkedAuthorsList } from 'app/containers/affiliation-form-dialog/linked-authors-list';
import { Person } from 'app/models/person';

interface AffiliationFormDialogProps {
  affiliation?: Affiliation;
  allowLinkAuthors?: boolean;
  onAccept: (affiliation: Affiliation, linkedAuthors: Person[]) => void;
  onCancel: () => void;
  onDelete: (affiliation: Affiliation) => void;
}

const labelProps = { shrink: true };

export const AffiliationFormDialog: React.FC<AffiliationFormDialogProps> = (props) => {
  const isNewAffiliation = !props.affiliation;
  const { allowLinkAuthors } = props;

  const affiliatedAuthors = useSelector(getAffiliatedAuthors)(props.affiliation?.id);
  const [userLinkedAuthors, setUserLinkedAuthors] = useState<Person[]>(affiliatedAuthors);
  const allAuthors = useSelector(getAuthors);

  const [affiliation, setAffiliation] = useState<Affiliation>(
    props.affiliation ||
      createAffiliation(undefined, {
        label: '',
        institution: {
          name: ''
        },
        address: {
          city: ''
        },
        country: ''
      })
  );
  const [isConfirmShown, setConfirmShow] = useState<boolean>(false);

  const classes = useAffiliationFormStyles();

  const handleDelete = useCallback(() => {
    setConfirmShow(true);
  }, [setConfirmShow]);

  const handleAcceptDeletion = useCallback(() => {
    setConfirmShow(false);
    props.onDelete(affiliation);
  }, [setConfirmShow, affiliation, props]);

  const handleRejectDeletion = useCallback(() => {
    setConfirmShow(false);
  }, [setConfirmShow]);

  const handleLinkedAuthorsChange = useCallback((authors: Person[]) => {
    setUserLinkedAuthors(authors);
  }, []);

  const handleFormChange = useCallback(
    (event: SyntheticEvent) => {
      const fieldName = event.target['name'];
      const newValue = event.target['value'];
      const newAffiliation = set(cloneDeep(affiliation), fieldName, newValue);
      setAffiliation(newAffiliation);
    },
    [affiliation, setAffiliation]
  );

  const handleDone = useCallback(() => {
    props.onAccept(affiliation, userLinkedAuthors);
  }, [props, affiliation, userLinkedAuthors]);

  return (
    <section className={classes.root}>
      <TextField
        fullWidth
        name="institution.name"
        label="Department/Institute"
        classes={{ root: classes.inputField }}
        InputLabelProps={labelProps}
        variant="outlined"
        multiline
        value={affiliation.institution.name}
        onChange={handleFormChange}
      />
      <TextField
        name="address.city"
        label="City"
        classes={{ root: classes.inputField }}
        InputLabelProps={labelProps}
        variant="outlined"
        value={affiliation.address.city}
        onChange={handleFormChange}
      />
      <TextField
        name="country"
        fullWidth
        label="Country"
        classes={{ root: classes.inputField }}
        InputLabelProps={labelProps}
        variant="outlined"
        value={affiliation.country}
        onChange={handleFormChange}
      />
      {allowLinkAuthors ? (
        <LinkedAuthorsList
          linkedAuthors={userLinkedAuthors}
          allAuthors={allAuthors}
          onChange={handleLinkedAuthorsChange}
        />
      ) : undefined}
      <div className={classes.buttonPanel}>
        {!isNewAffiliation ? (
          <ActionButton variant="outlinedWarning" onClick={handleDelete} title="Delete" />
        ) : undefined}
        <div aria-hidden={true} className={classes.spacer}></div>
        <ActionButton variant="secondaryOutlined" onClick={props.onCancel} title="Cancel" />
        <ActionButton variant="primaryContained" onClick={handleDone} title="Done" />
      </div>
      {isConfirmShown
        ? renderConfirmDialog(
            'You are deleting an affiliation',
            'Deleting an affiliation will remove all associated links with authors. Are you sure want to proceed?',
            handleAcceptDeletion,
            handleRejectDeletion
          )
        : undefined}
    </section>
  );
};
