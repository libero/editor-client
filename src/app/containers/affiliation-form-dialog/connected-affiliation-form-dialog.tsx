import React, { useCallback } from 'react';
import { Affiliation } from 'app/models/affiliation';
import { useDispatch } from 'react-redux';
import { AffiliationFormDialog } from 'app/containers/affiliation-form-dialog/affiliation-form-dialog';
import { Person } from 'app/models/person';
import * as manuscriptActions from 'app/actions/manuscript.actions';
import { linkAffiliationsAction } from 'app/actions/manuscript.actions';
import * as manuscriptEditorActions from 'app/actions/manuscript-editor.actions';

interface ConnectedAffiliationsFormDialogProps {
  affiliation: Affiliation;
}

export const ConnectedAffiliationsFormDialog: React.FC<ConnectedAffiliationsFormDialogProps> = ({ affiliation }) => {
  const dispatch = useDispatch();
  const isNewAffiliation = !affiliation;

  const closeDialog = useCallback(() => {
    dispatch(manuscriptEditorActions.hideModalDialog());
  }, [dispatch]);

  const handleAccept = useCallback(
    (affiliation: Affiliation, linkedAuthors: Person[]) => {
      if (isNewAffiliation) {
        dispatch(manuscriptActions.addAffiliationAction(affiliation));
      } else {
        dispatch(manuscriptActions.updateAffiliationAction(affiliation));
      }
      dispatch(linkAffiliationsAction({ affiliation, authors: linkedAuthors }));
      closeDialog();
    },
    [closeDialog, dispatch, isNewAffiliation]
  );

  const handleDelete = useCallback(
    (affiliation: Affiliation) => {
      dispatch(manuscriptActions.deleteAffiliationAction(affiliation));
      closeDialog();
    },
    [closeDialog, dispatch]
  );

  return (
    <AffiliationFormDialog
      affiliation={affiliation}
      onAccept={handleAccept}
      onDelete={handleDelete}
      onCancel={closeDialog}
    />
  );
};
