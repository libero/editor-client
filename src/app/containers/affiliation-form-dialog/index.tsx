import React, { SyntheticEvent, useCallback, useState } from 'react';
import { TextField } from '@material-ui/core';
import { useDispatch, useSelector } from 'react-redux';
import { set, cloneDeep } from 'lodash';

import { useAffiliationFormStyles } from './styles';
import * as manuscriptEditorActions from 'app/actions/manuscript-editor.actions';
import * as manuscriptActions from 'app/actions/manuscript.actions';
import { PromptDialog } from 'app/components/prompt-dialog';
import { ActionButton } from 'app/components/action-button';
import { Affiliation, createAffiliation } from 'app/models/affiliation';
import { getAffiliatedAuthors, getAuthors } from 'app/selectors/manuscript.selectors';
import { LinkedAuthorsList } from 'app/containers/affiliation-form-dialog/linked-authors-list';
import { Person } from 'app/models/person';
import { linkAffiliationsAction } from 'app/actions/manuscript.actions';

interface AffiliationFormDialogProps {
  affiliation?: Affiliation;
}

const renderConfirmDialog = (title: string, msg: string, onAccept: () => void, onReject: () => void) => {
  return (
    <PromptDialog
      title={title}
      message={msg}
      isOpen={true}
      onAccept={onAccept}
      onReject={onReject}
      acceptLabel="Delete"
      rejectLabel="Cancel"
      acceptVariant="containedWarning"
      rejectVariant="secondaryOutlined"
    />
  );
};

const labelProps = { shrink: true };

export const AffiliationFormDialog: React.FC<AffiliationFormDialogProps> = (props) => {
  const dispatch = useDispatch();
  const isNewAffiliation = !props.affiliation;

  const affiliatedAuthors = useSelector(getAffiliatedAuthors)(props.affiliation?.id);
  const [userLinkedAuthors, setUserLinkedAuthors] = useState<Person[]>(affiliatedAuthors);
  const allAuthors = useSelector(getAuthors);

  const [affiliation, setAffiliation] = useState<Affiliation>(
    props.affiliation ||
      createAffiliation(undefined, {
        label: '',
        institution: {
          department: '',
          name: ''
        },
        address: {
          city: ''
        },
        country: ''
      })
  );
  const [isConfirmShown, setConfirmShow] = useState<boolean>(false);

  const classes = useAffiliationFormStyles();

  const closeDialog = useCallback(() => {
    dispatch(manuscriptEditorActions.hideModalDialog());
  }, [dispatch]);

  const handleDelete = useCallback(() => {
    setConfirmShow(true);
  }, [setConfirmShow]);

  const handleAccept = useCallback(() => {
    setConfirmShow(false);
    dispatch(manuscriptActions.deleteAffiliationAction(affiliation));
    closeDialog();
  }, [setConfirmShow, affiliation, closeDialog, dispatch]);

  const handleReject = useCallback(() => {
    setConfirmShow(false);
  }, [setConfirmShow]);

  const handleLinkedAuthorsChange = useCallback((authors: Person[]) => {
    setUserLinkedAuthors(authors);
  }, []);

  const handleFormChange = useCallback(
    (event: SyntheticEvent) => {
      const fieldName = event.target['name'];
      const newValue = event.target['value'];
      const newAffiliation = set(cloneDeep(affiliation), fieldName, newValue);
      setAffiliation(newAffiliation);
    },
    [affiliation, setAffiliation]
  );

  const handleDone = useCallback(() => {
    if (isNewAffiliation) {
      dispatch(manuscriptActions.addAffiliationAction(affiliation));
    } else {
      dispatch(manuscriptActions.updateAffiliationAction(affiliation));
    }
    dispatch(linkAffiliationsAction({ affiliation, authors: userLinkedAuthors }));
    closeDialog();
  }, [affiliation, closeDialog, dispatch, isNewAffiliation, userLinkedAuthors]);

  return (
    <section className={classes.root}>
      <TextField
        fullWidth
        name="institution.department"
        label="Department"
        classes={{ root: classes.inputField }}
        InputLabelProps={labelProps}
        variant="outlined"
        value={affiliation.institution.department}
        onChange={handleFormChange}
      />
      <TextField
        fullWidth
        name="institution.name"
        label="Institute"
        classes={{ root: classes.inputField }}
        InputLabelProps={labelProps}
        variant="outlined"
        value={affiliation.institution.name}
        onChange={handleFormChange}
      />
      <TextField
        name="address.city"
        label="City"
        classes={{ root: classes.inputField }}
        InputLabelProps={labelProps}
        variant="outlined"
        value={affiliation.address.city}
        onChange={handleFormChange}
      />
      <TextField
        name="country"
        fullWidth
        label="Country"
        classes={{ root: classes.inputField }}
        InputLabelProps={labelProps}
        variant="outlined"
        value={affiliation.country}
        onChange={handleFormChange}
      />
      <LinkedAuthorsList
        linkedAuthors={userLinkedAuthors}
        allAuthors={allAuthors}
        onChange={handleLinkedAuthorsChange}
      />
      <div className={classes.buttonPanel}>
        {!isNewAffiliation ? (
          <ActionButton variant="outlinedWarning" onClick={handleDelete} title="Delete" />
        ) : undefined}
        <div aria-hidden={true} className={classes.spacer}></div>
        <ActionButton variant="secondaryOutlined" onClick={closeDialog} title="Cancel" />
        <ActionButton variant="primaryContained" onClick={handleDone} title="Done" />
      </div>
      {isConfirmShown
        ? renderConfirmDialog(
            'You are deleting an affiliation',
            'Deleting an affiliation will remove all associated links with authors. Are you sure want to proceed?',
            handleAccept,
            handleReject
          )
        : undefined}
    </section>
  );
};
