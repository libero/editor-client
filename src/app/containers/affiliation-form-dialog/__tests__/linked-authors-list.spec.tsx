import React from 'react';
import { Select, IconButton } from '@material-ui/core';
import { create, act } from 'react-test-renderer';
import { mount } from 'enzyme';

import { LinkedAuthorsList } from 'app/containers/affiliation-form-dialog/linked-authors-list';
import { Person } from 'app/models/person';
import { ActionButton } from 'app/components/action-button';

jest.mock('@material-ui/core', () => {
  return {
    Select: ({ children }) => <div data-cmp="Select">{children}</div>,
    MenuItem: ({ children }) => <div data-cmp="MenuItem">{children}</div>,
    FormControl: ({ children }) => <div data-cmp="FormControl">{children}</div>,
    InputLabel: ({ children }) => <div data-cmp="InputLabel">{children}</div>,
    IconButton: () => <div data-cmo="iconButton"></div>,
    Button: () => <div data-cmo="Button"></div>
  };
});

describe('Linked Authors List', () => {
  it('renders component', () => {
    const allAuthors = givenAuthors();
    const linkedAuthors = [allAuthors[0]];
    const wrapper = create(
      <LinkedAuthorsList linkedAuthors={linkedAuthors} allAuthors={allAuthors} onChange={jest.fn()} />
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('adds empty fields to the end of the list', () => {
    const allAuthors = givenAuthors();
    const clickHandler = jest.fn();
    const linkedAuthors = [allAuthors[0]];

    const wrapper = create(
      <LinkedAuthorsList linkedAuthors={linkedAuthors} allAuthors={allAuthors} onChange={clickHandler} />
    );

    act(() => {
      wrapper.root.findByType(ActionButton).props.onClick();
    });

    expect(clickHandler).not.toHaveBeenCalled();
    expect(wrapper).toMatchSnapshot();
  });

  it('removes elements from the list', () => {
    const allAuthors = givenAuthors();
    const clickHandler = jest.fn();
    const linkedAuthors = [allAuthors[0]];
    const wrapper = create(
      <LinkedAuthorsList linkedAuthors={linkedAuthors} allAuthors={allAuthors} onChange={clickHandler} />
    );

    act(() => {
      wrapper.root.findByType(ActionButton).props.onClick();
    });

    act(() => {
      wrapper.root.findByType(ActionButton).props.onClick();
    });

    expect(wrapper.root.findAllByType(Select).length).toBe(3);

    act(() => {
      wrapper.root.findAllByType(IconButton).pop().props.onClick();
    });

    expect(wrapper.root.findAllByType(Select).length).toBe(2);
    expect(wrapper).toMatchSnapshot();
    expect(clickHandler).toHaveBeenCalled();
  });

  it('fires an event when changes happen', () => {
    const allAuthors = givenAuthors();
    const clickHandler = jest.fn();
    const linkedAuthors = [allAuthors[0]];
    const wrapper = mount(
      <LinkedAuthorsList linkedAuthors={linkedAuthors} allAuthors={allAuthors} onChange={clickHandler} />
    );

    wrapper.find(ActionButton).prop('onClick')();
    wrapper.update();
    wrapper.find(Select).at(1).prop('onChange')({ target: { value: allAuthors[1].id } }, null);

    expect(clickHandler).toHaveBeenCalledWith([allAuthors[0], allAuthors[1]]);
  });
});

function givenAuthors(): Person[] {
  return [
    {
      firstName: 'Fred',
      lastName: 'Atherden',
      email: 'f.atherden@elifesciences.org',
      orcId: 'https://orcid.org/0000-0002-6048-1470',
      affiliations: ['aff1', 'aff2'],
      id: '149aa8ef-ed62-4ef9-aeaa-9758ce99c136'
    },
    {
      firstName: 'Jeanine',
      lastName: 'Smith',
      suffix: 'III',
      affiliations: ['aff2'],
      id: '8c62aed8-45c5-47fd-93bd-0816c3d2a110'
    },
    {
      firstName: 'Jack',
      lastName: 'London',
      suffix: 'III',
      affiliations: ['aff2'],
      id: '5a0204e2-5f4c-4947-9cbb-2f306d6702f9'
    },
    {
      firstName: 'Mark',
      lastName: 'Twain',
      suffix: 'III',
      affiliations: ['aff2'],
      id: '9da79de3-a9e2-47dd-bc1d-5c8c633c12f6'
    },
    {
      firstName: 'Alexandr',
      lastName: 'Solzhenitsin',
      suffix: 'III',
      affiliations: ['aff2'],
      id: '92f5e98a-e255-4c59-934c-fa174368e9ac'
    }
  ];
}
