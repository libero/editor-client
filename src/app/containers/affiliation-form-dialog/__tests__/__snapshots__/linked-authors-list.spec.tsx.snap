// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Linked Authors List adds empty fields to the end of the list 1`] = `
<div
  className="makeStyles-inputField-2"
>
  <div
    className="makeStyles-affiliatedAuthorRow-5"
  >
    <div
      data-cmp="FormControl"
    >
      <div
        data-cmp="InputLabel"
      >
        Affiliated Author
      </div>
      <div
        data-cmp="Select"
      >
        <div
          data-cmp="MenuItem"
        >
          None
        </div>
        <div
          data-cmp="MenuItem"
        >
          Fred Atherden
        </div>
        <div
          data-cmp="MenuItem"
        >
          Jeanine Smith III
        </div>
        <div
          data-cmp="MenuItem"
        >
          Jack London III
        </div>
        <div
          data-cmp="MenuItem"
        >
          Mark Twain III
        </div>
        <div
          data-cmp="MenuItem"
        >
          Alexandr Solzhenitsin III
        </div>
      </div>
    </div>
    <div
      data-cmo="iconButton"
    />
  </div>
  <div
    className="makeStyles-affiliatedAuthorRow-5"
  >
    <div
      data-cmp="FormControl"
    >
      <div
        data-cmp="InputLabel"
      >
        Affiliated Author
      </div>
      <div
        data-cmp="Select"
      >
        <div
          data-cmp="MenuItem"
        >
          None
        </div>
        <div
          data-cmp="MenuItem"
        >
          Jeanine Smith III
        </div>
        <div
          data-cmp="MenuItem"
        >
          Jack London III
        </div>
        <div
          data-cmp="MenuItem"
        >
          Mark Twain III
        </div>
        <div
          data-cmp="MenuItem"
        >
          Alexandr Solzhenitsin III
        </div>
      </div>
    </div>
    <div
      data-cmo="iconButton"
    />
  </div>
  <div
    data-cmo="Button"
  />
</div>
`;

exports[`Linked Authors List removes elements from the list 1`] = `
<div
  className="makeStyles-inputField-2"
>
  <div
    className="makeStyles-affiliatedAuthorRow-5"
  >
    <div
      data-cmp="FormControl"
    >
      <div
        data-cmp="InputLabel"
      >
        Affiliated Author
      </div>
      <div
        data-cmp="Select"
      >
        <div
          data-cmp="MenuItem"
        >
          None
        </div>
        <div
          data-cmp="MenuItem"
        >
          Fred Atherden
        </div>
        <div
          data-cmp="MenuItem"
        >
          Jeanine Smith III
        </div>
        <div
          data-cmp="MenuItem"
        >
          Jack London III
        </div>
        <div
          data-cmp="MenuItem"
        >
          Mark Twain III
        </div>
        <div
          data-cmp="MenuItem"
        >
          Alexandr Solzhenitsin III
        </div>
      </div>
    </div>
    <div
      data-cmo="iconButton"
    />
  </div>
  <div
    className="makeStyles-affiliatedAuthorRow-5"
  >
    <div
      data-cmp="FormControl"
    >
      <div
        data-cmp="InputLabel"
      >
        Affiliated Author
      </div>
      <div
        data-cmp="Select"
      >
        <div
          data-cmp="MenuItem"
        >
          None
        </div>
        <div
          data-cmp="MenuItem"
        >
          Jeanine Smith III
        </div>
        <div
          data-cmp="MenuItem"
        >
          Jack London III
        </div>
        <div
          data-cmp="MenuItem"
        >
          Mark Twain III
        </div>
        <div
          data-cmp="MenuItem"
        >
          Alexandr Solzhenitsin III
        </div>
      </div>
    </div>
    <div
      data-cmo="iconButton"
    />
  </div>
  <div
    data-cmo="Button"
  />
</div>
`;

exports[`Linked Authors List renders component 1`] = `
<div
  className="makeStyles-inputField-2"
>
  <div
    className="makeStyles-affiliatedAuthorRow-5"
  >
    <div
      data-cmp="FormControl"
    >
      <div
        data-cmp="InputLabel"
      >
        Affiliated Author
      </div>
      <div
        data-cmp="Select"
      >
        <div
          data-cmp="MenuItem"
        >
          None
        </div>
        <div
          data-cmp="MenuItem"
        >
          Fred Atherden
        </div>
        <div
          data-cmp="MenuItem"
        >
          Jeanine Smith III
        </div>
        <div
          data-cmp="MenuItem"
        >
          Jack London III
        </div>
        <div
          data-cmp="MenuItem"
        >
          Mark Twain III
        </div>
        <div
          data-cmp="MenuItem"
        >
          Alexandr Solzhenitsin III
        </div>
      </div>
    </div>
    <div
      data-cmo="iconButton"
    />
  </div>
  <div
    data-cmo="Button"
  />
</div>
`;
